## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [375]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [376]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [377]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [378]:
# Use this space to do any additional research
# and familiarize yourself with the data.



In [379]:
# What three item categories had the most entries?

catCounts = df["category"].value_counts()
print(catCounts.head(3))

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [380]:
# For the category with the most entries,
# which subcategory had the most entries?

subcatCounts = df[(df['category'] == 'consumables')]['subcategory'].value_counts()
print(subcatCounts.head(1))

subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [381]:
# Which five clients had the most entries in the data?

clientCounts = df['client_id'].value_counts()
print(clientCounts.head(5))

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [382]:
# Store the client ids of those top 5 clients in a list.

clientIDs = []
top5ClientIDs = []
for key, value in clientCounts.items():
    clientIDs.append(key)

top5ClientIDs = clientIDs[0:5]

print(top5ClientIDs)

[33615, 66037, 46820, 38378, 24741]


In [383]:
# How many total units (the qty column) did the
# client with the most entries order order?

# subcatCounts = df[(df['category'] == 'consumables')]['subcategory'].value_counts()

clientEntries = df[(df['client_id'] == top5ClientIDs[0])]

print(clientEntries['qty'].sum())

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [384]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['line_subtotal'] = round(df['unit_price'] * df['qty'], 2)

df[['unit_price', 'qty', 'line_subtotal']].head()

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00

In [385]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

import numpy as np

df['total_weight'] = round(df['unit_weight'] * df['qty'], 2)

df['shipping_price'] = np.where(df['total_weight'] > 50, df['total_weight'] * 7, df['total_weight'] * 10)

df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head()

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [386]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df['line_sales_tax'] = round((df['line_subtotal'] + df['shipping_price']) * 0.0925, 2)

df['line_price'] = round(df['line_subtotal'] + df['shipping_price'] + df['line_sales_tax'], 2)

df[['line_subtotal', 'shipping_price', 'line_price', 'line_sales_tax']].head()

line_subtotal  shipping_price  line_price  line_sales_tax
0      115164.00         5512.50   131839.08        11162.58
1         523.95          312.90      914.26           77.41
2         527.28          458.64     1077.12           91.20
3        1056.18          356.70     1543.57          130.69
4        3902.00         6500.20    11364.40          962.20

In [387]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['line_cost'] = round((df['unit_cost'] * df['qty']) + df['shipping_price'], 2)

df[['unit_cost', 'qty', 'shipping_price', 'line_cost']].head()

unit_cost  qty  shipping_price  line_cost
0     762.71  105         5512.50   85597.05
1      15.09   21          312.90     629.79
2       7.86   39          458.64     765.18
3      24.85   29          356.70    1077.35
4     108.17   20         6500.20    8663.60

In [388]:
# Create a column for the profit of
# each line using line cost and line price

df['line_profit'] = round(df['line_price'] - df['line_cost'], 2)

df[['line_price', 'line_cost', 'line_profit']].head()

line_price  line_cost  line_profit
0   131839.08   85597.05     46242.03
1      914.26     629.79       284.47
2     1077.12     765.18       311.94
3     1543.57    1077.35       466.22
4    11364.40    8663.60      2700.80

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [389]:
# Check your work using the totals above

onedf = df.query('order_id == 2742071')

total = onedf['line_price'].sum()

print(total) # 1

onedf = df.query('order_id == 2173913')

total = onedf['line_price'].sum()

print(total) # 2

onedf = df.query('order_id == 6128929')

total = onedf['line_price'].sum()

print(total) # 3

152811.88999999998
162388.71
923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [390]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

for id in top5ClientIDs:
    newdf = df.query(f"client_id == {id}")
    print(f"{id}: " + f"${round(newdf['line_price'].sum(), 2)}")

33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [414]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

newdf = pd.DataFrame()

for id in top5ClientIDs:
    qdf = df.query(f"client_id == {id}")

    newRow = {"client_id": id, 
              "qty": qdf['qty'].sum(), 
              "shipping_price": "$" + str(round(qdf['shipping_price'].sum() / 100000, 2)) + "M", 
              "line_price": "$" + str(round(qdf['line_price'].sum() / 100000, 2)) + "M", 
              "line_profit": "$" + str(round(qdf['line_profit'].sum() / 100000, 2)) + "M"}
    
    newdf = pd.concat([newdf, pd.DataFrame([newRow])], ignore_index=True)

newdf = newdf.sort_values(by=['line_profit'], ascending=False)

newdf.head()

client_id     qty shipping_price line_price line_profit
4      24741  239862        $51.26M   $822.69M     $365.8M
3      38378   73667        $34.29M   $129.07M     $32.72M
1      66037   43018        $13.95M    $102.6M     $32.55M
2      46820   75768        $16.01M    $97.44M     $27.37M
0      33615   64313        $18.29M    $83.77M     $22.02M

In [415]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

newdf = newdf.rename(columns={"client_id": "Client ID", "qty": "Units", "shipping_price": "Shipping", "line_price": "Total Revenue", "line_profit": "Total Profit"})

newdf.head()

Client ID   Units Shipping Total Revenue Total Profit
4      24741  239862  $51.26M      $822.69M      $365.8M
3      38378   73667  $34.29M      $129.07M      $32.72M
1      66037   43018  $13.95M       $102.6M      $32.55M
2      46820   75768  $16.01M       $97.44M      $27.37M
0      33615   64313  $18.29M       $83.77M      $22.02M

In [416]:
# Sort the updated data by "Total Profit" form highest to lowest
newdf.head()

Client ID   Units Shipping Total Revenue Total Profit
4      24741  239862  $51.26M      $822.69M      $365.8M
3      38378   73667  $34.29M      $129.07M      $32.72M
1      66037   43018  $13.95M       $102.6M      $32.55M
2      46820   75768  $16.01M       $97.44M      $27.37M
0      33615   64313  $18.29M       $83.77M      $22.02M